Frameworks/Foundation/NSOperation.mm

```diff
@@ -277,35 +296,37 @@ - (void)main {
  @Status Interoperable
 */
 - (void)waitUntilFinished {
-    pthread_mutex_lock(&priv->finishLock);
+    _finishLock.lock();
     while (![self isFinished]) {
-        pthread_cond_wait(&priv->finishCondition, &priv->finishLock);
+        _finishCondition.wait(_finishLock);
     }
-    pthread_mutex_unlock(&priv->finishLock);
+
+    _finishLock.unlock();
 }
 
 /**
  @Status Interoperable
 */
-- (id)dependencies {
-    return priv->dependencies;
+- (NSArray*)dependencies {
+    std::lock_guard<std::recursive_mutex> lock(_dependenciesLock);
+    NSArray* copy = [[_dependencies copy] autorelease];
+    return copy;
 }
 
 /**
  @Status Interoperable
 */
 - (void)dealloc {
-    assert(!priv->completionBlock);
-    delete priv;
-    [super dealloc];
-}
+    _dependenciesLock.lock();
+    int count = [_dependencies count];
+    for (int i = 0; i < count; i++) {
+        id op = [_dependencies objectAtIndex:i];
```

Comment:
This is likely better written as `for(NSOperation* op in _dependencies)`, as it can be optimized.